### PR TITLE
#163426522 Implement user signup validation

### PR DESCRIPTION
**Pull request by biolabalo**
#### What does this PR do?
- Implement validation  for user sign up
#### Description of Task to be completed?
- Users should receive descriptive signup/Registration validation errors
#### How should this be manually tested?
- After cloning the repo, cd into it and RUN `npm install` 

* Using postman, go to `/api/v1/register`
* Provide wrong values  into a json object with the following fields
  `email`
`username`
  `password`
#### What are the relevant pivotal tracker stories?
[#163426522](https://www.pivotaltracker.com/story/show/163426522)

### CODE REVIEW

**Review by EfeAgare — DISMISSED** [review]

**Review by divinediscipline — DISMISSED** [review]
Looks good to me. Nice work.

**Review by ozone4real — DISMISSED** [review]

File: server/tests/app/auth.test.js [hunk 1]
```diff
@@ -5,27 +5,80 @@ import app from '../../index';
...
-        expect(res.body.error.message).to.equal('Not Found');
+        expect(res.body.error).to.equal('route does not exist');
         expect(res.status).to.equal(404);
       });
   });
 });
 
 describe('user registration', () => {
-  it('should fail on registration, since the email field was not provided', () => request(app)
+  it('should fail on registration, since the email', () => request(app)
```
Comment (p-funky) [inline server/tests/app/auth.test.js]: Seems you mistakenly removed some of the test description.

File: server/tests/app/auth.test.js [hunk 2]
```diff
@@ -48,9 +101,91 @@ describe('user registration', () => {
...
+  it('should fail on registration, since the username is less than 3 characters', () => request(app)
+    .post('/api/v1/register')
+    .set('content-type', 'application/json')
+    .send(user5)
+    .then((res) => {
+      expect(res.status).to.equal(400);
+      expect(res.body.message).to.equal('Username must be at least 3 characters long');
+    }));
+
+  it('should fail on registration, since the password field field was not provided', () => request(app)
```
Comment (p-funky) [inline server/tests/app/auth.test.js]: The word "field" was repeated in the test description.

File: server/middlewares/validators/signUpValidator.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      errorMessage(res, 400, 'Username must be at least 3 characters long');
+      break;
+    case '"password" is not allowed to be empty':
+      errorMessage(res, 400, 'No password was specified');
+      break;
+    case '"password" length must be at least 8 characters long':
+      errorMessage(res, 400, 'Password length must be at least 8 characters long');
+      break;
+    case '"email" is not allowed to be empty':
+      errorMessage(res, 400, 'No Email was specified');
```
Comment (p-funky) [inline server/middlewares/validators/signUpValidator.js]: Message here should be "No email was specified". No need to capitalize the e in email.

File: server/middlewares/validators/signUpValidator.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import Joi from 'joi';
+import isAlphaNumeric from './isAplhaNumeric';
+import errorMessage from '../../helpers/errorHelpers';
+
+
+export default (req, res, next) => {
+  const { username, email, password } = req.body;
+
+  if (password && !isAlphaNumeric(password)) {
+    return errorMessage(res, 400, 'Password Should be Alphanumeric');
```
Comment (p-funky) [inline server/middlewares/validators/signUpValidator.js]: Message here should be "Password should be Alphanumeric".

File: server/middlewares/validators/signUpValidator.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  const { error } = Joi.validate({ username, password, email, }, schema);
+
+  if (!error) return next();
+
+  const errorMessageFromJoi = error.details[0].message;
+
+  switch (errorMessageFromJoi) {
+    case '"username" is not allowed to be empty':
+      errorMessage(res, 400, 'No Username was specified');
```
Comment (p-funky) [inline server/middlewares/validators/signUpValidator.js]: Message here should be "No username was specified".

**Comment by p-funky** [comment]
Some tests would need some minor updating once you've made the requested changes.

**Review by bobbylemm — DISMISSED** [review]
LGTM

**Review by omenkish — DISMISSED** [review]

**Review by p-funky — APPROVED** [review]

**Review by primuse — APPROVED** [review]

**Pull request merged**